### PR TITLE
CAAS: Adds add-caas command, and AddCloud API.

### PR DESCRIPTION
**Pull request by mikemccracken**
## Description of change
This is all behind the "CAAS" feature flag.

Adds 'add-caas' command to add a cloud definition for a CAAS cluster.
We will reuse the Cloud type in state to store endpoints & creds for CAAS clusters.

- Adds "AddCloud" to Cloud facade. Revs Cloud version to 2 when CAAS feature flag is set
Traditional clouds are added via add-cloud before a controller is bootstrapped, and stored locally until bootstrap, at which point they're passed in via the bootstrap initialization data, and added directly (not using the API) by the controller during init. CAAS clouds are added to a pre-existing controller, so we needed to expose AddCloud in the Cloud facade.

What's missing, to be added in future branches:
- doesn't read CAAS config (e.g. $KUBECONFIG contents) to get creds or endpoint yet - plan is to add a clientconfig package to read that info
- no validation of given cloud type - plan is to add a CAASProvider to be the CAAS version of Environ
- needs an interactive version of the command (eventually)

## QA steps
with a bootstrapped controller: `juju add-caas anything somename`, then `juju clouds` should list it, and a cloud should show up in mongodb on the controller also, e.g.:

```
juju:PRIMARY> db.clouds.find()
{ "_id" : "localhost", "name" : "localhost", "type" : "lxd", "auth-types" : [ "interactive", "certificate" ], "endpoint" : ... etc ... , "txn-queue" : [ ] }
{ "_id" : "kubernetes", "name" : "kubernetes", "type" : "myk8s2", "auth-types" : [ "userpass" ], "endpoint" : "", "txn-revno" : NumberLong(2), "txn-queue" : [ "596eb0bb6792ac2655c9dbc3_9473df14" ] }
```

## Documentation changes

This will definitely affect user workflow, but is currently hidden behind a feature flag and the specific UX is up for discussion.

### CODE REVIEW

File: cmd/juju/caas/add.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	PersonalCloudMetadata() (map[string]cloud.Cloud, error)
+	WritePersonalCloudMetadata(cloudsMap map[string]cloud.Cloud) error
+}
+
+// Implemented by cloudapi.Client
+type AddCloudAPI interface {
+	AddCloud(cloud.Cloud) error
+}
+
+var usageAddCaasSummary = `
```
Comment (mikemccracken) [inline cmd/juju/caas/add.go]: ⚠️  jsing pointed out that this isn't proper go naming. Needs to be usageAddCAASSummary. Lots of other examples in the branch, will fix.

File: api/cloud/cloud_test.go [hunk 2]
```diff
@@ -286,3 +286,58 @@ func (s *cloudSuite) TestCredentials(c *gc.C) {
...
+	client := cloudapi.NewClient(apiCaller)
+	err := client.AddCloud(cloud.Cloud{
+		Name:      "foo",
+		Type:      "dummy",
+		AuthTypes: []cloud.AuthType{cloud.EmptyAuthType, cloud.UserPassAuthType},
+		Regions:   []cloud.Region{{Name: "nether", Endpoint: "endpoint"}},
+	})
+
+	c.Assert(err, jc.ErrorIsNil)
+}
```
Comment (mjs) [inline api/cloud/cloud_test.go]: These tests are good except that they don't prove that the test apiCaller was actually called. Perhaps set a flag inside the dummy APICallerFunc and check that it's set afterwards?

File: apiserver/params/cloud.go [hunk 3]
```diff
@@ -107,3 +117,52 @@ type CloudSpecResult struct {
...
+	return jujucloud.Cloud{
+		Name:             cloudName,
+		Type:             p.Type,
+		AuthTypes:        authTypes,
+		Endpoint:         p.Endpoint,
+		IdentityEndpoint: p.IdentityEndpoint,
+		StorageEndpoint:  p.StorageEndpoint,
+		Regions:          regions,
+	}
+}
```
Comment (mjs) [inline apiserver/params/cloud.go]: I don't think these 2 helper functions should live here. The `params` package is imported by both `apiserver/...` and `api/...` so should have minimal dependencies of its own (this change adds the `jujucloud` dep).

Somewhere under `apiserver/common` might be more appropriate.

Comment (mikemccracken) [inline apiserver/params/cloud.go]: Thanks, I was wondering about where they belonged.

**Comment by mikemccracken** [comment]
$$merge$$

**Pull request merged**